static/app/components/tours/tourContext.tsx

```diff
@@ -0,0 +1,157 @@
+import {type Dispatch, type Reducer, useCallback, useMemo, useRef} from 'react';
+import {useReducer} from 'react';
+
+export type TourEnumType = string | number;
+
+export interface TourStep<T extends TourEnumType> {
+  /**
+   * The element to focus on when the tour step is active.
+   * This is usually set within the TourElement component, which wraps the focused element.
+   */
+  element: HTMLElement | null;
+  /**
+   * Unique ID for the tour step.
+   */
+  id: T;
+}
+
+type TourStartAction<T extends TourEnumType> = {
+  type: 'START_TOUR';
+  stepId?: T;
+};
+type TourNextStepAction = {
+  type: 'NEXT_STEP';
+};
+type TourPreviousStepAction = {
+  type: 'PREVIOUS_STEP';
+};
+type TourEndAction = {
+  type: 'END_TOUR';
+};
+
+export type TourAction<T extends TourEnumType> =
+  | TourStartAction<T>
+  | TourNextStepAction
+  | TourPreviousStepAction
+  | TourEndAction;
+
+export interface TourState<T extends TourEnumType> {
+  /**
+   * The current active tour step. If this is null, the tour is not active.
+   */
+  currentStep: TourStep<T> | null;
+  /**
+   * Whether the tour is available to the user. Should be set by flags or other conditions.
+   */
+  isAvailable: boolean;
+  /**
+   * The ordered step IDs. Declared once when the provider is initialized.
+   */
+  orderedStepIds: readonly T[];
+}
+
+type TourRegistry<T extends TourEnumType> = {
+  [key in T]: TourStep<T> | null;
+};
+
+export function useTourReducer<T extends TourEnumType>(
+  initialState: TourState<T>
+): TourContextType<T> {
+  const {orderedStepIds} = initialState;
+  const registry = useRef<TourRegistry<T>>({} as TourRegistry<T>);
+  const isCompletelyRegistered = Object.values(registry.current).every(Boolean);
```

Comment:
I don’t think this works reliably. Since `registry` is a `ref`, `isCompletelyRegistered` will not change when something is added to / removed from the registry. That means the reducer will always see the initial value, unless the context provider itself re-renders (which can happen for other reasons).

Also, `isCompletelyRegistered` starts out as `true`, as the registry starts with an empty object, and `Object.values({}).every(Boolean)` is `true`.

Reply:
Great catch! It's intentional to use a ref to avoid the re-renders when a new component registers ([per comment](https://github.com/getsentry/sentry/pull/85900#discussion_r1972339450)),  but this boolean shouldn't be declared outside of reducer, it should check the ref every time it's called. Though thinking about it some more this is probably not pure anymore so I will figure out an alternative.

Reply:
Okay it should be pure now!